# CLASSIFYING THE CLASSES OF WINE FROM THE WINE DATASET

# Data Preprocessing

In [1]:
import pandas as pd
df= pd.read_csv('wine.csv')
df

1  14.23  1.71  2.43  15.6  127   2.8  3.06   .28  2.29   5.64  1.04  \
0    1  13.20  1.78  2.14  11.2  100  2.65  2.76  0.26  1.28   4.38  1.05   
1    1  13.16  2.36  2.67  18.6  101  2.80  3.24  0.30  2.81   5.68  1.03   
2    1  14.37  1.95  2.50  16.8  113  3.85  3.49  0.24  2.18   7.80  0.86   
3    1  13.24  2.59  2.87  21.0  118  2.80  2.69  0.39  1.82   4.32  1.04   
4    1  14.20  1.76  2.45  15.2  112  3.27  3.39  0.34  1.97   6.75  1.05   
..  ..    ...   ...   ...   ...  ...   ...   ...   ...   ...    ...   ...   
172  3  13.71  5.65  2.45  20.5   95  1.68  0.61  0.52  1.06   7.70  0.64   
173  3  13.40  3.91  2.48  23.0  102  1.80  0.75  0.43  1.41   7.30  0.70   
174  3  13.27  4.28  2.26  20.0  120  1.59  0.69  0.43  1.35  10.20  0.59   
175  3  13.17  2.59  2.37  20.0  120  1.65  0.68  0.53  1.46   9.30  0.60   
176  3  14.13  4.10  2.74  24.5   96  2.05  0.76  0.56  1.35   9.20  0.61   

     3.92  1065  
0    3.40  1050  
1    3.17  1185  
2    3.45  1480  
3    2.93   735  
4    2.85  1450  
..    ...   ...  
172  1.74   740  
173  1.56   750  
174  1.56   835  
175  1.62   840  
176  1.60   560  

[177 rows x 14 columns]

In [13]:
df.columns=['class','Alcohol','Malicadid','Ash','Alcalinityofash','Magnesium','Totalphenols','Flavanoids','nonflavanoidphenols','Proanthocyanins','Colorintensity','Hue','OD280/OD315','Proline']
df

class  Alcohol  Malicadid   Ash  Alcalinityofash  Magnesium  \
0        1    13.20       1.78  2.14             11.2        100   
1        1    13.16       2.36  2.67             18.6        101   
2        1    14.37       1.95  2.50             16.8        113   
3        1    13.24       2.59  2.87             21.0        118   
4        1    14.20       1.76  2.45             15.2        112   
..     ...      ...        ...   ...              ...        ...   
172      3    13.71       5.65  2.45             20.5         95   
173      3    13.40       3.91  2.48             23.0        102   
174      3    13.27       4.28  2.26             20.0        120   
175      3    13.17       2.59  2.37             20.0        120   
176      3    14.13       4.10  2.74             24.5         96   

     Totalphenols  Flavanoids  nonflavanoidphenols  Proanthocyanins  \
0            2.65        2.76                 0.26             1.28   
1            2.80        3.24                 0.30             2.81   
2            3.85        3.49                 0.24             2.18   
3            2.80        2.69                 0.39             1.82   
4            3.27        3.39                 0.34             1.97   
..            ...         ...                  ...              ...   
172          1.68        0.61                 0.52             1.06   
173          1.80        0.75                 0.43             1.41   
174          1.59        0.69                 0.43             1.35   
175          1.65        0.68                 0.53             1.46   
176          2.05        0.76                 0.56             1.35   

     Colorintensity   Hue  OD280/OD315  Proline  
0              4.38  1.05         3.40     1050  
1              5.68  1.03         3.17     1185  
2              7.80  0.86         3.45     1480  
3              4.32  1.04         2.93      735  
4              6.75  1.05         2.85     1450  
..              ...   ...          ...      ...  
172            7.70  0.64         1.74      740  
173            7.30  0.70         1.56      750  
174           10.20  0.59         1.56      835  
175            9.30  0.60         1.62      840  
176            9.20  0.61         1.60      560  

[177 rows x 14 columns]

In [14]:
#finding mean, minimum, maximum values to find outliers
df.describe()

class     Alcohol   Malicadid         Ash  Alcalinityofash  \
count  177.000000  177.000000  177.000000  177.000000       177.000000   
mean     1.943503   12.993672    2.339887    2.366158        19.516949   
std      0.773991    0.808808    1.119314    0.275080         3.336071   
min      1.000000   11.030000    0.740000    1.360000        10.600000   
25%      1.000000   12.360000    1.600000    2.210000        17.200000   
50%      2.000000   13.050000    1.870000    2.360000        19.500000   
75%      3.000000   13.670000    3.100000    2.560000        21.500000   
max      3.000000   14.830000    5.800000    3.230000        30.000000   

        Magnesium  Totalphenols  Flavanoids  nonflavanoidphenols  \
count  177.000000    177.000000  177.000000           177.000000   
mean    99.587571      2.292260    2.023446             0.362316   
std     14.174018      0.626465    0.998658             0.124653   
min     70.000000      0.980000    0.340000             0.130000   
25%     88.000000      1.740000    1.200000             0.270000   
50%     98.000000      2.350000    2.130000             0.340000   
75%    107.000000      2.800000    2.860000             0.440000   
max    162.000000      3.880000    5.080000             0.660000   

       Proanthocyanins  Colorintensity         Hue  OD280/OD315      Proline  
count       177.000000      177.000000  177.000000   177.000000   177.000000  
mean          1.586949        5.054802    0.956983     2.604294   745.096045  
std           0.571545        2.324446    0.229135     0.705103   314.884046  
min           0.410000        1.280000    0.480000     1.270000   278.000000  
25%           1.250000        3.210000    0.780000     1.930000   500.000000  
50%           1.550000        4.680000    0.960000     2.780000   672.000000  
75%           1.950000        6.200000    1.120000     3.170000   985.000000  
max           3.580000       13.000000    1.710000     4.000000  1680.000000

##### There are no outliers in the dataset as all values in the respective columns lie close to the mean value of the columns

# Scaling the Data

In [10]:
from sklearn.preprocessing import MinMaxScaler
scaler=MinMaxScaler()
scaler.fit(df)

MinMaxScaler()

##### Scaled the dataset using MinMaxScaler

# Decision tree

In [18]:
from sklearn.tree import DecisionTreeClassifier
#converting pandas dataframe to arrays
X=df[df.columns[1:13]]
y=df["class"]
clf= DecisionTreeClassifier()
clf=clf.fit(X,y)

In [28]:
from sklearn.model_selection import train_test_split
from sklearn.metrics import accuracy_score
X_train, X_test, y_train, y_test = train_test_split(X,y,test_size=0.25, random_state = 100)

In [29]:
y_pred=clf.predict(X_test)
print("Decision tree accuracy is",accuracy_score(y_test, y_pred)*100)

Decision tree accuracy is 100.0


In [30]:
y_pred

array([1, 2, 1, 2, 3, 3, 2, 3, 1, 2, 1, 3, 3, 2, 2, 1, 3, 1, 1, 3, 2, 2,
       3, 1, 3, 3, 2, 3, 3, 3, 1, 2, 2, 3, 3, 3, 2, 2, 3, 1, 2, 3, 2, 2,
       2], dtype=int64)

In [31]:
#using 10 fold cross validation
from sklearn.model_selection import cross_val_score
print(cross_val_score(clf, X,y, cv=10))
Decision_tree_score=cross_val_score(clf,X,y, cv=10)

[0.88888889 0.88888889 1.         0.72222222 0.88888889 1.
 0.94444444 0.94117647 0.94117647 1.        ]


In [33]:
import statistics
statistics.mean(Decision_tree_score)*100

91.63398692810458

##### Accuracy of Decision tree using 10 fold cross validation is 91.63398692810458

# Random forest

In [36]:
from sklearn.ensemble import RandomForestClassifier

In [41]:
randomforest=RandomForestClassifier(n_estimators=10)

In [42]:
randomforest.fit(X,y)

RandomForestClassifier(n_estimators=10)

In [43]:
y_pred=randomforest.predict(X_test)

In [44]:
print('Random Forest accuracy is',accuracy_score(y_test, y_pred)*100)

Random Forest accuracy is 100.0


In [45]:
from sklearn.model_selection import cross_val_score
print(cross_val_score(clf, X,y, cv=10))
Random_Forest_score=cross_val_score(clf,X,y, cv=10)

[0.83333333 0.77777778 0.94444444 0.72222222 0.88888889 1.
 0.94444444 1.         0.94117647 1.        ]


In [47]:
statistics.mean(Random_Forest_score)*100

91.07843137254902

##### The accuracy of Random Forest using 10 fold cross validation is 91.07843137254902

# KNN Classifier

In [54]:
from sklearn.neighbors import KNeighborsClassifier
kneighbor=KNeighborsClassifier(n_neighbors=10)

In [59]:
K_Neighbor_score=cross_val_score(kneighbor,X,y, cv=10)

In [60]:
statistics.mean(K_Neighbor_score)*100

75.84967320261438

In [62]:
for k in range(1,50):
    kneighbor= KNeighborsClassifier(n_neighbors=k)
    K_Neighbor_score=cross_val_score(kneighbor,X,y, cv=10)
    print(statistics.mean(K_Neighbor_score)*100)

85.3921568627451
80.35947712418302
86.47058823529412
81.40522875816994
82.51633986928104
79.15032679738562
80.29411764705883
79.77124183006536
77.54901960784314
75.84967320261438
77.02614379084967
76.99346405228759
76.40522875816993
77.54901960784314
77.54901960784314
78.0718954248366
78.0718954248366
75.81699346405229
72.41830065359477
73.56209150326798
74.18300653594771
72.48366013071895
73.0718954248366
73.0392156862745
73.0718954248366
71.37254901960785
71.89542483660131
71.33986928104575
70.22875816993464
70.19607843137254
67.90849673202615
67.94117647058823
65.65359477124183
65.65359477124183
64.50980392156863
65.06535947712419
65.09803921568627
65.09803921568627
64.54248366013073
65.09803921568627
65.13071895424837
63.39869281045751
63.39869281045751
63.39869281045751
63.98692810457517
64.54248366013073
63.39869281045751
62.84313725490196
62.84313725490196


##### Accuracy of KNN is changing with change in K.Value of K means a substantial difference in KNN model.Accuracy of KNN is high when K=3Accuracy of KNN is 86.47058823529412


# NAIVE BAYES

# MultinomialNB

In [64]:
from sklearn.naive_bayes import MultinomialNB

In [67]:
MultinomialNB_classifier=MultinomialNB()
MultinomialNB_classifier_scores=cross_val_score(MultinomialNB_classifier, X,y,cv=10)
statistics.mean(MultinomialNB_classifier_scores)*100

86.66666666666667

##### Accuracy of MultinomialNB with 10 fold cross validation is 86.66666666666667

# GaussianNB

In [68]:
from sklearn.naive_bayes import GaussianNB

In [69]:
GaussianNB_classifier= GaussianNB()
GaussianNB_classifier_scores=cross_val_score(GaussianNB_classifier, X,y,cv=10)
statistics.mean(GaussianNB_classifier_scores)*100

94.41176470588235

##### Accuracy of GaussianNB classifier with 10 fold cross validation is 94.41176470588235

### Conclusion: All the models are classifying well. Among them GaussianNB model is having the high accuracy score of 94.41176470588235 and the least accuracy score is MultinomialNB with 86.66666666666667 